### DISCUSSION TITLE
Enable shader alpha(transparency) blend mode

### DISCUSSION THREAD

**Discussion by salah-walid** [iTwin/itwinjs-core#3408] [Q&A]
Hi,

I added a fragment shader that takes the texture and change alpha channel according to a uniform variable, i have looked around and found that to enable opacity in GLSL you have to enable a special blend mode with the following line:
`glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);`

But the shader compiler is not recognizing GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA
Here is my fragment shader:
`vec4 effectMain() {`
 `         glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);`
`          vec4 color = TEXTURE(u_diffuse, v_texCoord);`
 `         return vec4(color.rgb, u_opacity);`
`}`

**Comment by pmconne** [comment]
`glBlendFunc` is an OpenGL API; the corresponding WebGL API is `WebGLRenderingContext.blendFunc`.
These APIs are called from C++/JavaScript. They are not accessible in shader code.

A screen-space effect only mutates an input image. It has nothing to blend with, so it doesn't make sense to change the blend mode.
Can you explain what your intended effect should do? For example, if you set u_opacity to 0 (fully transparent), what do you expect to be displayed in the viewport?

**Reply by salah-walid** [reply]
if opacity is 0 it should make the BIM and the terrain invisible(ui should still be visible)

**Reply by pmconne** [reply]
So you want to blend the viewport's contents with the HTML content underneath?
That can [currently be done](https://www.itwinjs.org/learning/display/timelineanimation/) for the viewport's **background** pixels. It doesn't apply to the geometry within the viewport.
If your goal is to apply a uniform opacity to the canvas itself, why not do that using CSS instead of trying to use a screen-space effect?
(If that's **not** your goal, please specify what your goal is).

**Reply by salah-walid** [reply]
Does the canvas include the side menu widgets and navigation buttons ? if not is there a class name i can target
What i want is for a way to change the geometry's opacity without changing the opacity of the UI elements

**Reply by pmconne** [reply]
In display-test-app:
```
viewport.vpDiv.style.opacity = "0.5";
```
![image](https://user-images.githubusercontent.com/22944042/159902419-69791bc6-7aa6-4f28-a932-24b809b4e6c1.png)

The UI framework doesn't add its controls to the viewport's div so they should be unaffected.

**Reply by pmconne** [reply]
The classname is "imodeljs-vp".
